pkg/kv/kvserver/client_merge_test.go

```diff
@@ -3407,7 +3407,9 @@ func testMergeWatcher(t *testing.T, injectFailures bool) {
 						roachpb.NewTransactionRetryError(roachpb.RETRY_SERIALIZABLE, "filter err"))
 				}
 			}
-			if pt := req.GetPushTxn(); pt != nil {
+			// We can detect PushTxn requests generated by the watcher goroutine
+			// because they use the minimum transaction priority.
+			if pt := req.GetPushTxn(); pt != nil && pt.PusherTxn.Priority == enginepb.MinTxnPriority {
```

Comment:
Thanks for looking into this! Is there a way to sharpen this further by checking the `pt.PusheeTxn.Key` matches that start key of the range at which the txn is anchored (I think that's the LHS)?

Reply:
So there's this thing, we need a key '/Local/Range/System/"tse"/RangeDescriptor' we can probably generate that upfront, but that would rely on range for test key "a" not changing. Alternatively, we can look on range descriptor, but it is not available until the server is started and we have range. That would require pushing this value to filter closure.

Reply:
I just looked at this test and this is all a bit outdated. The test uses keys "a", "b", "c" but it should really use something closer to what's provided by `append(append([]byte(nil), tc.ScratchRange()...), "a")` etc. to use less ad-hoc keys. ScratchRange is static: https://github.com/cockroachdb/cockroach/blob/9cc036d4e3013e4bbdfdf620410339843eb6ce85/pkg/server/testserver.go#L1225-L1228
and I've had multiple situations in which I really just wanted to hard-code it in the test (which I think is completely reasonable). So perhaps we can add (in package `keys`) `var ScratchRangeKey = TableDataMax` and adjust the snippet above to use that.
With all that done, you should be able to target the particular key the transaction is anchored at as `RangeDescriptorKey(keys.ScratchRangeKey and "a")` or some version thereof, without having to inject some shared memory in the filter.

If you'd rather not go down that rabbit hole, that's ok and the current fix LGTM, though the above is what I'd suggest if you have the stomach for it.

Reply:
I really like that because I think relying on priority is sketchy and is fragile. I might also change the other test from which I took the priority approach.

Reply:
So it doesn't work with filtering on range descriptor key because split transactions use the same keys. I changed tests to use scratch ranges anyways as it seems as a better approach. But to make transaction more selective we can just use name, because those names are used consistently by watcher. They are even declared as private constants, maybe it is worth exporting them for the sake of tests?

Reply:
Sounds good! I am also okay (maybe slightly prefer) leaving the name unexported and adding a comment that tests rely on it (i.e. do a string search before you change it). We're unlikely to ever want to change it.